src/System.Private.DataContractSerialization/src/System/Runtime/Serialization/Json/ReflectionJsonFormatReader.cs

```diff
@@ -0,0 +1,216 @@
+            Debug.Assert(jsonContext != null);
+
+            bool canReadSimpleDictionary = collectionContract.Kind == CollectionKind.Dictionary
+                                        || collectionContract.Kind == CollectionKind.GenericDictionary;
+            if (canReadSimpleDictionary && jsonContext.UseSimpleDictionaryFormat)
+            {
+                ReadSimpleDictionary(xmlReader, context, collectionContract, collectionContract.ItemType, resultCollection);
+            }
+
+            return false;
+        }
+
+        private void ReadSimpleDictionary(XmlReaderDelegator xmlReader, XmlObjectSerializerReadContext context, CollectionDataContract collectionContract, Type keyValueType, object dictionary)
+        {
+            Type[] keyValueTypes = keyValueType.GetGenericArguments();
+            Type keyType = keyValueTypes[0];
+            Type valueType = keyValueTypes[1];
+
+            int keyTypeNullableDepth = 0;
+            Type keyTypeOriginal = keyType;
+            while (keyType.GetTypeInfo().IsGenericType && keyType.GetGenericTypeDefinition() == Globals.TypeOfNullable)
+            {
+                keyTypeNullableDepth++;
+                keyType = keyType.GetGenericArguments()[0];
+            }
+
+            ClassDataContract keyValueDataContract = (ClassDataContract)collectionContract.ItemContract;
+            DataContract keyDataContract = keyValueDataContract.Members[0].MemberTypeContract;
+
+            KeyParseMode keyParseMode = KeyParseMode.Fail;
+
+            if (keyType == Globals.TypeOfString || keyType == Globals.TypeOfObject)
+            {
+                keyParseMode = KeyParseMode.AsString;
+            }
+            else if (keyType.GetTypeInfo().IsEnum)
+            {
+                keyParseMode = KeyParseMode.UsingParseEnum;
+            }
+            else if (keyDataContract.ParseMethod != null)
+            {
+                keyParseMode = KeyParseMode.UsingCustomParse;
+            }
+
+            if (keyParseMode == KeyParseMode.Fail)
+            {
+                throw DiagnosticUtility.ExceptionUtility.ThrowHelperError(
+                    new InvalidOperationException(
+                        SR.Format(SR.KeyTypeCannotBeParsedInSimpleDictionary,
+                            DataContract.GetClrTypeFullName(collectionContract.UnderlyingType),
+                            DataContract.GetClrTypeFullName(keyType))
+                    ));
+            }
+
+            while (true)
+            {
+                XmlNodeType nodeType = xmlReader.MoveToContent();
+                if (nodeType == XmlNodeType.EndElement)
+                {
+                    return;
+                }
+                if (nodeType != XmlNodeType.Element)
+                {
+                    throw XmlObjectSerializerReadContext.CreateUnexpectedStateException(XmlNodeType.Element, xmlReader);
+                }
+
+                context.IncrementItemCount(1);
```

Comment:
I agree. In most (if not all) cases we increment the count by 1. But I will leave this as is for now as it's what we have been using.